pyxem/signals/indexation_results.py

```diff
@@ -121,7 +121,73 @@ def _get_best_match(z):
         array with shape (5,)
 
     """
-    return z[np.argmax(z[:, -1]),:] 
+    return z[np.argmax(z[:, -1]), :]
+
+
+def results_dict_to_crystal_map(results, phase_key_dict, diffraction_library=None):
```

Comment:
I checked out the branch locally and varied navigation shape (1D or 2D), n_best (1 or more), and number of phases (1 or 2). Too things needs improving:
1. There are too many assumptions on data shape
2. When `results` is multi phase only one rotation per point should be passed to CrystalMap: CrystalMap supports only one phase ID per point, but multiple rotations per point when the map is single phase.

Here's a working variant of the function which accomodates all these changes and adds a few more things. Feel free to take this and commit it yourself:

```python
def results_dict_to_crystal_map(results, phase_key_dict, diffraction_library=None):
    """Export an indexation result from
    :func:`index_dataset_with_template_rotation` to a crystal map with
    `n_best` rotations, score, mirrors and one phase ID per data point.
    
    Parameters
    ----------
    results : dict
        Results dictionary obtained from
        :func:`index_dataset_with_template_rotation`.
    phase_key_dict : dict
        Dictionary mapping phase ID to phase name, obtained from
        :func:`index_dataset_with_template_rotation`.
    diffraction_library : diffsims.libraries.DiffractionLibrary, optional
        Used for the structures to be passed to
        :class:`orix.crystal_map.PhaseList`.

    Returns
    -------
    orix.crystal_map.CrystalMap
        Crystal map containing `results`. The map has multiple rotations
        and properties ("correlation", "mirrored_template",
        "template_index") per point only if `n_best` passed to
        :func:`index_dataset_with_template_rotation` were more than one
        and "phase_index" only has one phase.

    Notes
    -----
    Phase's :attr:`~orix.crystal_map.Phase.point_group` must be set
    manually to the correct :class:`~orix.quaternion.Symmetry` after
    the crystal map is returned.
    """
    nx, ny, n_best = results["phase_index"].shape
    n_points = nx * ny

    # Best matching phase
    phase_id = results["phase_index"][..., 0].ravel()
    n_phases = np.unique(phase_id).size

    x, y = np.indices((nx, ny)).reshape((2, n_points))

    if diffraction_library is not None:
        structures = diffraction_library.structures
    else:
        structures = None
    phase_list = PhaseList(names=phase_key_dict, structures=structures)

    euler = np.deg2rad(results["orientation"].reshape((n_points, n_best, 3)))
    if n_phases > 1:
        euler = euler[:, 0]  # Keep best match only
    euler = euler.squeeze()  # Remove singleton dimensions
    rotations = Rotation.from_euler(euler, convention="bunge", direction="crystal2lab")

    props = {}
    for key in ("correlation", "mirrored_template", "template_index"):
        try:
            prop = results[key]
            if n_phases > 1:
                prop = prop[..., 0].ravel()  # Keep best match only
            else:
                prop = prop.reshape((n_points, n_best))
            props[key] = prop.squeeze()  # Remove singleton dimensions
        except KeyError:
            pass

    return CrystalMap(
        rotations=rotations,
        phase_id=phase_id,
        x=x,
        y=y,
        phase_list=phase_list,
        prop=props,
    )

```

Reply:
Yeah ok, I think this does make more sense.  I wasn't sure what should be done if it was multiphase so was just keeping all the rotations but using only the best phase.

Reply:
In that case, for a map with multiple rotations per point, the 2nd, 3rd and so on best matching `CrystalMap.orientations` in a point could be assigned the incorrect point group if the best matching phase IDs varied in `results["phase_index"]` for that point. The approach above ensures that only one rotation and properties per point are returned if `results["phase_index"]` has more than one phase.

Reply:
```python
    for key in ("correlation", "mirrored_template", "template_index"):
        try:
            prop = results[key]
            if n_phases > 1:
                prop = prop[..., 0].ravel()  # Keep best match only
            else:
                prop = prop.reshape((n_points, n_best))
            props[key] = prop.squeeze()  # Remove singleton dimensions
        except KeyError:
            pass
```

Why do you think that passing by not having these in the dictionary is better than it failing, just stability?

Reply:
Good question, I should have said why I added it. Actually, I would rewrite it as

```python
import warnings
[...]
    for key in ("correlation", "mirrored_template", "template_index"):
        try:
            prop = results[key]
        except KeyError:
            warnings.warn(f"Property '{key}' not found in `results`")
        if n_phases > 1:
            prop = prop[..., 0].ravel()  # Keep best match only
        else:
            prop = prop.reshape((n_points, n_best))
        props[key] = prop.squeeze()  # Remove singleton dimensions
```

It shouldn't be a requirement to have these exact dictionary keys in `results` in order to return a crystal map from it: only rotations are strictly required. I've had to add these exceptions to other code previously because the initial implementation was too restrictive.

Reply:
Yeah ok, with a warning I think it makes more sense, and because these aren't required for the xmap the code similarly shouldn't require it.